dwst/scripts/particles.js

```diff
@@ -20,7 +20,37 @@ const specialChars = [
   '\\',
 ];
 
-export class InvalidParticles extends Error { }
+export class InvalidParticles extends Error {
+  constructor(remainder, expected, expectedLengthOverride) {
+    super();
+    this.expression = null;
+    this.errorPosition = null;
+    this.remainder = remainder;
+    this.expected = expected;
+    if (typeof expectedLengthOverride === 'undefined') {
+      this.expected_length = expected.length;
+    } else {
+      this.expected_length = expectedLengthOverride;
+    }
+    this.got = remainder.substring(0, this.expected_length);
+  }
+}
+
+export function invalidToMlog(invalidParticle) {
+  return [
+    'Invalid template.',
+    invalidParticle.expression,
+    `${' '.repeat(invalidParticle.errorPosition)}^`,
+    `Expected ${invalidParticle.expected} but got ${invalidParticle.got} instead.`,
+  ];
+}
+
+export function finishInvalidParticle(invalidParticle, particleString) {
+  invalidParticle.expression = particleString;
+  const errorPosition = particleString.length - invalidParticle.remainder.length + invalidParticle.expected_length - 1;
```

Comment:
The length of the expected thing should not matter if the indicator arrow is always at the first incorrect character. Am I missing something here?

Reply:
Yes but a special case is the beginning of a particle.
Consider broken particle ```$foo(bar)}```.
Right now our code thinks anything starting with ```$``` is an instruction particle (see particles.js: ```function readParticle```).
But the first thing the parser then tries to do is skip ```${```. (see particles.js: ```function skipExpressionOpen```).
Our broken particle will at that point make the function throw a new InvalidParticles Error.
The error position will be calculated comparing what was parsed and what remains to be parsed. 
But our broken particle has made the parser fail before anything got accepted as parsed, and all of the particle string remains considered unparsed. 
So the string difference will cause the error position to point at index 0, which would be the character ```$``` instead of ```f```.

I think I was trying to fix the wrong thing. We should work on skipExpressionOpen.

Reply:
Maybe. I'm also curious if our tests are somehow broken or not ideal for this special case. Perhaps adding the new test cases for problematic cases will make this easier to implement.